### PR TITLE
Report parser name and location in XContent deprecation warnings

### PR DESCRIPTION
**Pull request by romseygeek**
It's simple to deprecate a field used in an `ObjectParser` just by adding deprecation
markers to the relevant `ParseField` objects.  The warnings themselves don't currently
have any context - they simply say that a deprecated field has been used, but not
*where* in the input xcontent it appears.  This commit adds the parent object parser
name and `XContentLocation` to these deprecation messages.

### CODE REVIEW

**Comment by romseygeek** [comment]
`ParseField.match()` is used in nearly 1,000 places in the code outside of an ObjectParser, so this PR leaves the existing method signature and adds another one, with a BWC layer so that if we don't have the parent parser name then we emit the old version of the error message.   We can use the calls to the old signature to find places to convert to using `ObjectParser`.

I also changed the 'this field has been removed entirely' message to read more easily.

File: server/src/main/java/org/elasticsearch/common/xcontent/LoggingDeprecationHandler.java [hunk 1]
```diff
@@ -49,17 +51,33 @@ private LoggingDeprecationHandler() {
...
     @Override
-    public void usedDeprecatedName(String usedName, String modernName) {
-        deprecationLogger.deprecated("Deprecated field [{}] used, expected [{}] instead", usedName, modernName);
+    public void usedDeprecatedName(String parserName, Supplier<XContentLocation> location, String usedName, String modernName) {
+        if (parserName != null) {
+            deprecationLogger.deprecated("[{}][{}] Deprecated field [{}] used, expected [{}] instead",
+                parserName, location.get(), usedName, modernName);
+        } else {
+            deprecationLogger.deprecated("Deprecated field [{}] used, expected [{}] instead", usedName, modernName);
+        }
```
Comment (pugnascotia) [inline server/src/main/java/org/elasticsearch/common/xcontent/LoggingDeprecationHandler.java]: The only suggestion I'd make is about whether it's possible to consolidate the log messages. For example:

```suggestion
        String prefix = parserName == null ? "" :
            "[" + parserName + "][" + location.get() + "] ";
        deprecationLogger.deprecated("{}Deprecated field [{}] used, expected [{}] instead", prefix, usedName, modernName);
```

**Review by pugnascotia — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/common/xcontent/LoggingDeprecationHandler.java [hunk 2]
```diff
@@ -52,30 +52,24 @@ private LoggingDeprecationHandler() {
...
+            prefix, usedName, replacedWith);
     }
 
     @Override
     public void usedDeprecatedField(String parserName, Supplier<XContentLocation> location, String usedName) {
-        if (parserName != null) {
-            deprecationLogger.deprecated("[{}][{}] Deprecated field [{}] used, this field is unused and will be removed entirely",
-                parserName, location.get(), usedName);
+        String prefix = parserName == null ? "" : "[" + parserName + "][" + location.get() + "] ";
+        if (true) {
```
Comment (pugnascotia) [inline server/src/main/java/org/elasticsearch/common/xcontent/LoggingDeprecationHandler.java]: This doesn't look right 🤔

Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/common/xcontent/LoggingDeprecationHandler.java]: It wasn't!

Comment (pugnascotia) [reply server/src/main/java/org/elasticsearch/common/xcontent/LoggingDeprecationHandler.java]: Sorry, out-of-date diff 😅

**Review by pugnascotia — APPROVED** [review]

**Comment by romseygeek** [comment]
Thanks for the review @pugnascotia !

**Pull request merged**